signer-service/src/api/services/alchemypay.service.js

```diff
@@ -0,0 +1,196 @@
+const crypto = require('crypto');
+const { quoteProviders } = require('../../config/vars');
+
+function apiSign(timestamp, method, requestUrl, body, secretKey) {
+  const content = timestamp + method.toUpperCase() + getPath(requestUrl) + getJsonBody(body);
+  return crypto.createHmac('sha256', secretKey).update(content).digest('base64');
+}
+
+function getPath(requestUrl) {
+  const uri = new URL(requestUrl);
+  const path = uri.pathname;
+  const params = Array.from(uri.searchParams.entries());
+
+  if (params.length === 0) {
+    return path;
+  } else {
+    const sortedParams = [...params].sort(([aKey], [bKey]) => aKey.localeCompare(bKey));
+    const queryString = sortedParams.map(([key, value]) => `${key}=${value}`).join('&');
+    return `${path}?${queryString}`;
+  }
+}
+
+function getJsonBody(body) {
+  let map;
+
+  try {
+    map = JSON.parse(body);
+  } catch (error) {
+    map = {};
+    console.error("Couldn't parse JSON body", error);
+  }
+
+  if (Object.keys(map).length === 0) {
+    return '';
+  }
+
+  map = removeEmptyKeys(map);
+  map = sortObject(map);
+
+  return JSON.stringify(map);
+}
+
+function removeEmptyKeys(map) {
+  const retMap = {};
+
+  for (const [key, value] of Object.entries(map)) {
+    if (value !== null && value !== '') {
+      retMap[key] = value;
+    }
+  }
+
+  return retMap;
+}
+
+function sortObject(obj) {
+  if (typeof obj === 'object') {
+    if (Array.isArray(obj)) {
+      return sortList(obj);
+    } else {
+      return sortMap(obj);
+    }
+  }
+
+  return obj;
+}
+
+function sortMap(map) {
+  const sortedMap = new Map(Object.entries(removeEmptyKeys(map)).sort(([aKey], [bKey]) => aKey.localeCompare(bKey)));
+
+  for (const [key, value] of sortedMap.entries()) {
+    if (typeof value === 'object') {
+      sortedMap.set(key, sortObject(value));
+    }
+  }
+
+  return Object.fromEntries(sortedMap.entries());
+}
+
+function sortList(list) {
```

Comment:
All the code is adopted from the code snippet for JavaScript you can find on [this](https://alchemypay.readme.io/docs/api-sign) page. Since the AlchemyPay API seems not very forgiving to me, I think it's best if we do exactly like they do in the snippet.